daemon/graphdriver/overlay2/overlay.go

```diff
@@ -592,7 +596,20 @@ func (d *Driver) Get(id, mountLabel string) (_ containerfs.ContainerFS, retErr e
 	for i, s := range splitLowers {
 		absLowers[i] = path.Join(d.home, s)
 	}
-	opts := indexOff + "lowerdir=" + strings.Join(absLowers, ":") + ",upperdir=" + diffDir + ",workdir=" + workDir
+	var readonly bool
+	if _, err := os.Stat(path.Join(dir, "committed")); err == nil {
+		readonly = true
+	} else if !os.IsNotExist(err) {
+		return nil, err
+	}
+
+	var opts string
+	if readonly {
+		opts = indexOff + "lowerdir=" + diffDir + ":" + strings.Join(absLowers, ":")
+	} else {
+		opts = indexOff + "lowerdir=" + strings.Join(absLowers, ":") + ",upperdir=" + diffDir + ",workdir=" + workDir
+	}
```

Comment:
nit: I generally prefer something like this to remove the repetition.

```go
	opts := indexOff + "lowerdir=" + diffDir + ":" + strings.Join(absLowers, ":")
	if !readonly {
		opts += ",upperdir=" + diffDir + ",workdir=" + workDir
	}
```

Reply:
`diffDir` is only in lower for the readonly case

Reply:
Ah right, i missed diffDir.